thirdparty/nginx/nginx-1.7.1-zimbra/src/http/modules/ngx_http_upstream_zmauth_module.c

```diff
@@ -1043,6 +1054,174 @@ ngx_field_from_zmauthtoken(ngx_log_t *log, ngx_pool_t *pool,
+            "zmauth: ngx_get_authheader_bearer entered");
+
+    ngx_str_t authValue;
+    ngx_str_t token;
+
+    authValue = headers_in->authorization->value;
+
+    ngx_log_debug1 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth: ngx_get_authheader_bearer token:%V",&authValue);
+
+    if (authValue.len >= sizeof("Bearer ") - 1
+            && ngx_memcmp(authValue.data,"Bearer ",sizeof("Bearer ")-1)
+            == 0)
+    {
+        ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth: ngx_get_authheader_bearer Bearer token found");
+
+        token = authValue;
+        token.data += (sizeof("Bearer ") - 1);
+        token.len -= (sizeof("Bearer ") - 1);
+
+        ngx_log_debug1 (NGX_LOG_DEBUG_HTTP, log, 0, "zmauth: ngx_get_authheader_bearer Bearer token found:%V",&token);
+        *value = token;
+
+        return 1; 
+    }
+
+    ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0,
+            "zmauth: ngx_get_authheader_bearer exit");
+
+    return 0;
+}
+
+/* extract claim data from jwt token */
+static ngx_flag_t
+ngx_claimdata_from_jwttoken(ngx_log_t *log, ngx_pool_t *pool,
+        ngx_str_t *authtoken, ngx_str_t *field, ngx_str_t *value) {
+
+    ngx_flag_t f;
+    ngx_log_debug0 (NGX_LOG_DEBUG_HTTP, log, 0,
+            "zmauth: ngx_claimdata_from_jwttoken entered");
+
+    ngx_str_t pattern = ngx_string("(?<=\\.)(.*?)(?=\\.)");
+    ngx_regex_compile_t rgc;
+    u_char errstr[NGX_MAX_CONF_ERRSTR];
+    ngx_memzero(&rgc, sizeof(ngx_regex_compile_t));
+
+    rgc.pattern  = pattern;
+    rgc.pool     = pool;
+    rgc.err.len  = NGX_MAX_CONF_ERRSTR;
+    rgc.err.data = errstr;
+
+
+    if (ngx_regex_compile(&rgc) != NGX_OK)
+    {
+        ngx_log_error(NGX_LOG_INFO, log, 0,"zmauth: ngx_claimdata_from_jwttoken regex compilation failed.Error=====%s",errstr);
+        return 0;
+    }
+
+    int captures[30];
+    ngx_int_t  n;
+
+    n = ngx_regex_exec(rgc.regex, authtoken, captures, 30);
+    if (n >= 0)
+    {
+        /* string matches expression */
+        ngx_log_debug0(NGX_LOG_DEBUG_ZIMBRA, log, 0, "zmauth: ngx_claimdata_from_jwttoken match found");
+        ngx_str_t encodedPayload;
+        encodedPayload.data = authtoken->data + captures[0];
+        encodedPayload.len = captures[1] - captures[0];
+        ngx_log_debug1(NGX_LOG_DEBUG_ZIMBRA, log, 0, "zmauth: ngx_claimdata_from_jwttoken payload is %V",&encodedPayload);
+
+        ngx_str_t urlDecodedPayload;
+        if (ngx_decode_base64url(&urlDecodedPayload, &encodedPayload) != NGX_OK)
+        {
+            ngx_log_error(NGX_LOG_INFO, log, 0,"zmauth: ngx_claimdata_from_jwttoken base64url decoding failed");
+            return 0;
+        }
+        ngx_log_debug1(NGX_LOG_DEBUG_ZIMBRA, log, 0, "zmauth: ngx_claimdata_from_jwttoken base64url decoded payload is %V",&urlDecodedPayload);
+
+        int ret = ngx_claimdata_from_payload(log,pool,&urlDecodedPayload,value);
+        return ret; 
+    }
+    else if (n == NGX_REGEX_NO_MATCHED)
+    {
+        /* no match was found */
```

Comment:
nit: remove superfluous comment (here and in block below, possibly even in the block above)